File: ts/src/feature/test/TestFeature.ts

import type { Context, FeatureOptions } from '../../types'
import type { UniversalSDK } from '../../UniversalSDK'

import { BaseFeature } from '../base/BaseFeature'


const S_NOT_FOUND = 'Not found'


class TestFeature extends BaseFeature {
  version = '0.0.1'
  name = 'test'
  active = true

  _client?: UniversalSDK
  _options?: any


  init(ctx: Context, options: FeatureOptions): void | Promise<any> {
    this._client = ctx.client
    this._options = options

    const struct = ctx.utility.struct
    const walk = struct.walk
    const size = struct.size
    const setprop = struct.setprop

    const entity = this._options.entity

    this._client._mode = 'test'

    // Ensure entity ids are correct.
    walk(entity, (k: any, v: any, _parent: any, path: any) => {
      if (2 === size(path)) {
        setprop(v, 'id', k)
      }
      return v
    })

    const self = this

    function testFetcher(ctx: any, _fullurl: string, _fetchdef: any) {
      const struct = ctx.utility.struct
      const param = ctx.utility.param

      const getprop = struct.getprop
      const clone = struct.clone
      const merge = struct.merge
      const getelem = struct.getelem
      const select = struct.select
      const delprop = struct.delprop
      const getdef = struct.getdef

      function respond(status: number, data?: any, res?: any) {
        const out = merge([
          {
            status,
            statusText: 'OK',
            json: async () => data,
            body: 'not-used',
          },
          getdef(res, {})
        ])

        const headers: any = getprop(out, 'headers', {})

        // JS specific iterator.
        out.headers = {
          forEach(callback: any) {
            Object.keys(headers).forEach((key) => {
              callback(headers[key], key, this)
            })
          }
        }

        return out
      }


      const op = ctx.op
      const entmap = getprop(entity, op.entity, {})

      if ('load' === op.name) {
        const args = self.buildArgs(ctx, op, ctx.reqmatch)
        const found = select(entmap, args)
        const ent = getelem(found, 0)
        if (null == ent) {
          return respond(404, undefined, { statusText: S_NOT_FOUND })
        }
        else {
          delprop(ent, '$KEY')
          const out = clone(ent)
          return respond(200, out)
        }
      }
      else if ('list' === op.name) {
        const args = self.buildArgs(ctx, op, ctx.reqmatch)
        const found = select(entmap, args)
        if (null == found) {
          return respond(404, undefined, { statusText: S_NOT_FOUND })
        }
        else {
          found.map((ent: any) => delprop(ent, '$KEY'))
          const out = clone(found)
          return respond(200, out)
        }
      }
      else if ('update' === op.name) {
        const args = self.buildArgs(ctx, op, ctx.reqdata)
        const found = select(entmap, args)
        const ent = getelem(found, 0)
        if (null == ent) {
          return respond(404, undefined, { statusText: S_NOT_FOUND })
        }
        else {
          merge([ent, (ctx.reqdata || {})])
          delprop(ent, '$KEY')
          const out = clone(ent)
          return respond(200, out)
        }
      }
      else if ('remove' === op.name) {
        const args = self.buildArgs(ctx, op, ctx.reqmatch)
        const found = select(entmap, args)
        const ent = getelem(found, 0)
        if (null == ent) {
          return respond(404, undefined, { statusText: S_NOT_FOUND })
        }
        else {
          delprop(entmap, getprop(ent, 'id'))
          return respond(200)
        }
      }
      else if ('create' === op.name) {
        const args = self.buildArgs(ctx, op, ctx.reqdata)
        let id = param(ctx, 'id')
        if (null == id) {
          id = ((1e4 * Math.random() | 0).toString(16) +
            (1e4 * Math.random() | 0).toString(16) +
            (1e4 * Math.random() | 0).toString(16) +
            (1e4 * Math.random() | 0).toString(16)).padEnd(16, '0')
        }

        const ent = clone(ctx.reqdata)
        setprop(ent, 'id', id)
        setprop(entmap, id, ent)
        delprop(ent, '$KEY')
        const out = clone(ent)
        return respond(200, out)
      }
    }

    ctx.utility.fetcher = testFetcher
  }


  buildArgs(ctx: any, op: any, args: any): any {
    const struct = ctx.utility.struct
    const param = ctx.utility.param

    const getprop = struct.getprop
    const keysof = struct.keysof
    const getpath = struct.getpath
    const getelem = struct.getelem
    const select = struct.select
    const transform = struct.transform
    const isempty = struct.isempty

    const opname = getprop(op, 'name')
    const target =
      getelem(getpath(ctx.config, [
        'entity', getprop(ctx.entity, 'name'), 'op', opname, 'targets']), -1)

    const reqd = transform(
      select(getpath(target, ['args', 'params']), { reqd: true }),
      ['`$EACH`', '', '`$KEY.name`']
    )

    const qand: any[] = []
    const q = { '`$AND`': qand }

    for (let k of keysof(args)) {
      if ('id' === k || !isempty(select(reqd, k))) {
        const v = param(ctx, k)
        const ka = getprop(op.alias, k)

        let qor: any = [{ [k]: v }]
        if (null != ka) {
          qor.push({ [ka]: v })
        }

        qor = { '`$OR`': qor }

        qand.push(qor)
      }
    }

    if (ctx.ctrl.explain) {
      ctx.ctrl.explain.test = { query: q }
    }

    return q
  }
}


export {
  TestFeature
}

File: go/feature/test_feature.go
package feature

import (
	"fmt"
	"math/rand"

	vs "github.com/voxgig/struct"

	"voxgiguniversalsdk/core"
)

type TestFeature struct {
	BaseFeature
	client  *core.UniversalSDK
	options map[string]any
}

func NewTestFeature() *TestFeature {
	return &TestFeature{
		BaseFeature: BaseFeature{
			Version: "0.0.1",
			Name:    "test",
			Active:  true,
		},
	}
}

func (f *TestFeature) Init(ctx *core.Context, options map[string]any) {
	f.client = ctx.Client
	f.options = options

	entity := core.ToMapAny(vs.GetProp(options, "entity"))

	f.client.Mode = "test"

	// Ensure entity ids are correct.
	vs.Walk(entity, func(key *string, val any, parent any, path []string) any {
		if len(path) == 2 {
			if m, ok := val.(map[string]any); ok {
				if key != nil {
					m["id"] = *key
				}
			}
		}
		return val
	})

	self := f

	testFetcher := func(ctx *core.Context, _fullurl string, _fetchdef map[string]any) (any, error) {
		respond := func(status int, data any, extra map[string]any) map[string]any {
			out := map[string]any{
				"status":     status,
				"statusText": "OK",
				"json":       (func() any)(func() any { return data }),
				"body":       "not-used",
			}
			if extra != nil {
				for k, v := range extra {
					out[k] = v
				}
			}
			return out
		}

		op := ctx.Op
		entmap := core.ToMapAny(vs.GetProp(entity, op.Entity))
		if entmap == nil {
			entmap = map[string]any{}
		}

		if op.Name == "load" {
			args := self.buildArgs(ctx, op, ctx.Reqmatch)
			found := vs.Select(entmap, args)
			ent := vs.GetElem(found, 0)
			if ent == nil {
				return respond(404, nil, map[string]any{"statusText": "Not found"}), nil
			}
			vs.DelProp(ent, "$KEY")
			out := vs.Clone(ent)
			return respond(200, out, nil), nil
		} else if op.Name == "list" {
			args := self.buildArgs(ctx, op, ctx.Reqmatch)
			found := vs.Select(entmap, args)
			if found == nil {
				return respond(404, nil, map[string]any{"statusText": "Not found"}), nil
			}
			for _, item := range found {
				vs.DelProp(item, "$KEY")
			}
			out := vs.Clone(found)
			return respond(200, out, nil), nil
		} else if op.Name == "update" {
			args := self.buildArgs(ctx, op, ctx.Reqdata)
			found := vs.Select(entmap, args)
			ent := vs.GetElem(found, 0)
			if ent == nil {
				return respond(404, nil, map[string]any{"statusText": "Not found"}), nil
			}
			if entm, ok := ent.(map[string]any); ok {
				reqdata := ctx.Reqdata
				if reqdata != nil {
					for k, v := range reqdata {
						entm[k] = v
					}
				}
			}
			vs.DelProp(ent, "$KEY")
			out := vs.Clone(ent)
			return respond(200, out, nil), nil
		} else if op.Name == "remove" {
			args := self.buildArgs(ctx, op, ctx.Reqmatch)
			found := vs.Select(entmap, args)
			ent := vs.GetElem(found, 0)
			if ent == nil {
				return respond(404, nil, map[string]any{"statusText": "Not found"}), nil
			}
			if entm, ok := ent.(map[string]any); ok {
				id := vs.GetProp(entm, "id")
				vs.DelProp(entmap, id)
			}
			return respond(200, nil, nil), nil
		} else if op.Name == "create" {
			_ = self.buildArgs(ctx, op, ctx.Reqdata)
			id := ctx.Utility.Param(ctx, "id")
			if id == nil {
				id = fmt.Sprintf("%04x%04x%04x%04x",
					rand.Intn(0x10000), rand.Intn(0x10000),
					rand.Intn(0x10000), rand.Intn(0x10000))
			}

			ent := vs.Clone(ctx.Reqdata)
			if entm, ok := ent.(map[string]any); ok {
				entm["id"] = id
				if idStr, ok := id.(string); ok {
					entmap[idStr] = entm
				}
				vs.DelProp(entm, "$KEY")
				out := vs.Clone(entm)
				return respond(200, out, nil), nil
			}
			return respond(200, ent, nil), nil
		}

		return respond(404, nil, map[string]any{"statusText": "Unknown operation"}), nil
	}

	ctx.Utility.Fetcher = testFetcher
}

func (f *TestFeature) buildArgs(ctx *core.Context, op *core.Operation, args map[string]any) any {
	opname := op.Name

	// Get last target from config.
	targets := vs.GetPath([]any{"entity", ctx.Entity.GetName(), "op", opname, "targets"}, ctx.Config)
	target := vs.GetElem(targets, -1)

	// Get required params.
	paramsPath := vs.GetPath([]any{"args", "params"}, target)
	reqdParams := vs.Select(paramsPath, map[string]any{"reqd": true})
	reqd := vs.Transform(reqdParams, []any{"`$EACH`", "", "`$KEY.name`"})

	qand := []any{}
	q := map[string]any{"`$AND`": &qand}

	if args != nil {
		for _, key := range vs.KeysOf(args) {
			isId := key == "id"
			selected := vs.Select(reqd, key)
			isReqd := !vs.IsEmpty(selected)

			if isId || isReqd {
				v := ctx.Utility.Param(ctx, key)
				ka := vs.GetProp(op.Alias, key)

				qor := []any{map[string]any{key: v}}
				if ka != nil {
					if kas, ok := ka.(string); ok {
						qor = append(qor, map[string]any{kas: v})
					}
				}

				qand = append(qand, map[string]any{"`$OR`": qor})
			}
		}
	}

	// Update the slice behind the pointer.
	q["`$AND`"] = qand

	if ctx.Ctrl.Explain != nil {
		ctx.Ctrl.Explain["test"] = map[string]any{"query": q}
	}

	return q
}
